### PR TITLE
196 holdings and availability updates

### PR DESCRIPTION
**Pull request by banukutlu**
Requires psu-libraries/psulib_traject#162 and psu-libraries/psulib_traject#166

Updates mentioned in Issue #196:
-  Use currentLocationID instead of homeLocationID
-  Display fields: Removed Status, added Material column:https://github.com/psu-libraries/psulib_blacklight/wiki/Holdings-and-Availability-Display#display-fields
-  https://github.com/psu-libraries/psulib_blacklight/issues/196#issuecomment-502157816
-  Archival Thesis availability https://github.com/psu-libraries/psulib_blacklight/wiki/Holdings-and-Availability-Display#request-digitization-of-archival-thesis 
-  Updated error msg: https://github.com/psu-libraries/psulib_blacklight/wiki/Holdings-and-Availability-Display#error-message-when-webservices-not-available
- Use `title_245ab_tsim` instead of `title_display_ssm` for Illiad and Aeon urls

Extra work: fixes I want It button not displaying when there is only checkedout items - mentioned in #341

### CODE REVIEW

File: app/javascript/psulib_blacklight/javascripts/availability.js [hunk 1]
```diff
@@ -109,47 +117,68 @@ function loadAvailability(locations) {
                 // Now that the availability data has been rendered, check for ILL options and update links
                 $('.availability-holdings [data-type="ill-link"]').each(function() {
                     var catkey = $(this).data('catkey');
-                    createILLURL($(this), catkey)
+                    var callNumber = $(this).data('call-number');
+                    var archivalThesis = $(this).is('[data-archival-thesis]');
+                    var item = {
+                        catkey: catkey,
+                        callNumber: callNumber,
+                        archivalThesis: archivalThesis
```
Comment (cdmo) [inline app/javascript/psulib_blacklight/javascripts/availability.js]: Did you swap the `item` object's contents? ILL URL has `archivalThesis`. Seems like that should be handed to createAeonURL. ?

File: app/javascript/psulib_blacklight/javascripts/availability.js [hunk 2]
```diff
@@ -109,47 +117,68 @@ function loadAvailability(locations) {
...
+                        itemLocation: itemLocation,
+                        itemID: itemID
+                    };
+                    createAeonURL($(this), item);
                 });
             }).fail(function(data) {
                 $('.availability').each(function () {
                     $(this).addClass('availability-error alert alert-light');
-                    $(this).html("Error: could not determine availability");
+                    $(this).html("Please check back shortly for item availability or <a href=\"https://libraries.psu.edu/ask\">ask a librarian</a> for assistance.");
```
Comment (cdmo) [inline app/javascript/psulib_blacklight/javascripts/availability.js]: 👍

File: app/javascript/psulib_blacklight/javascripts/availability.js [hunk 3]
```diff
@@ -228,50 +257,73 @@ function groupByLibrary(holdings) {
...
+        location = `<a data-type="ill-link" data-catkey="${item.catkey}" data-call-number="${item.callNumber}" href="#">${spinner}Copy unavailable, request via Interlibrary Loan</a>`;
+    } else if (['ARKTHESES', 'AH-X-TRANS'].includes(item.currentLocationID)) {
+        var aeonLocation = (item.currentLocationID in all_locations) ? all_locations[item.currentLocationID] : "";
+        var shared = `data-catkey="${item.catkey}" data-call-number="${item.callNumber}" data-archival-thesis`;
+
+        location = `<a data-type="ill-link" ${shared} href="#">${spinner}Request Scan - Penn State Users</a><br>
+                    <a href="https://psu.illiad.oclc.org/illiad/upm/lending/lendinglogon.html">Request Scan - Guest</a><br>
+                    <a data-type="aeon-link" ${shared} data-item-id="${item.itemID}" data-item-location="${aeonLocation}" href="#">${spinner}View in Special Collections</a>`;
+    } else {
+        location = (item.currentLocationID in all_locations) ? all_locations[item.currentLocationID] : "";
```
Comment (cdmo) [inline app/javascript/psulib_blacklight/javascripts/availability.js]: This pattern shows up a couple of times. Would be nice to have a generic function to say "map this string to the value of the key that matches that string on this object or return an empty string. Call it "mapLocation" then you could read the code easier and eliminate a nested if. Like

`location = mapLocation(currentLocationID, mapObject)`

Comment (cdmo) [reply app/javascript/psulib_blacklight/javascripts/availability.js]: I meant this as a question

File: app/javascript/psulib_blacklight/javascripts/availability.js [hunk 4]
```diff
@@ -228,50 +257,73 @@ function groupByLibrary(holdings) {
...
-                statusCode = statusText = 'Missing';
-                break;
-            case 'ILLEND':
-                statusCode = 'ILLEND';
-                statusText = `<a data-type="ill-link" data-catkey="${titleID}" href="#"><span class="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span> This copy unavailable, submit request via Interlibrary Loan</a>`;
-                break;
-            default:
-                statusCode = statusText = 'Not Available';
-        }
+function resolveLocation(item, request_via_ill_locations, all_locations) {
```
Comment (cdmo) [inline app/javascript/psulib_blacklight/javascripts/availability.js]:
```suggestion
function printLocationHTML(item, request_via_ill_locations, all_locations) {
```

??

File: app/javascript/psulib_blacklight/javascripts/availability.js [hunk 5]
```diff
@@ -228,50 +257,73 @@ function groupByLibrary(holdings) {
...
     }
 
-    return {statusCode:statusCode, statusText: statusText};
+    return location;
 }
 
-function createILLURL(jQueryObj, catkey) {
-    $.get(`/catalog/${catkey}/raw.json`, function(data) {
-        var ILLURL = "https://psu-illiad-oclc-org.ezaccess.libraries.psu.edu/illiad/upm/illiad.dll?Action=10&Form=30";
+function createILLURL(jQueryObj, item) {
```
Comment (cdmo) [inline app/javascript/psulib_blacklight/javascripts/availability.js]: Really like how you've modified var names for consistency and readability.

File: app/javascript/psulib_blacklight/javascripts/availability.js (hunk 1 above)
Comment (cdmo) [reply app/javascript/psulib_blacklight/javascripts/availability.js]: nvm, I see that it is needed.

File: app/javascript/psulib_blacklight/javascripts/availability.js [hunk 6]
```diff
@@ -228,50 +257,73 @@ function groupByLibrary(holdings) {
...
+        var ILLURL = "https://psu-illiad-oclc-org.ezaccess.libraries.psu.edu/illiad/upm/illiad.dll/OpenURL?Action=10";
+
         if (Object.keys(data).length > 0) {
-            var ISBN = data.isbn_ssm;
-            var title = data.title_display_ssm;
-            var author = data.author_tsim;
-            var pubdate = data.pub_date_itsi;
-            ILLURL += `&isbn=${ISBN}&title=${title}&aulast=${author}.&rfr_id=info%3Asid%2Fcatalog.libraries.psu.edu&date=%5B${pubdate}%5D`;
+            var title = data.title_245ab_tsim;
+            var author = data.author_tsim ? data.author_tsim : "";
```
Comment (cdmo) [inline app/javascript/psulib_blacklight/javascripts/availability.js]: I guess this is necessary eh? If `data.author_tsim` is missing does it break things?

Comment (banukutlu) [reply app/javascript/psulib_blacklight/javascripts/availability.js]: yeah it was passing it as `undefined`

File: app/javascript/psulib_blacklight/javascripts/availability.js [hunk 7]
```diff
@@ -21,10 +22,12 @@ $(document).on('turbolinks:load', function() {
 /**
  * Load real time holdings and availability info from Sirsi Web Services
  */
-function loadAvailability(locations) {
+function loadAvailability(locations, item_types) {
     // Load Sirsi locations
     var all_locations = locations.locations;
     var all_libraries = locations.libraries;
+    var request_via_ill_locations = locations.request_via_ill;
+    var all_item_types = item_types.item_types;
```
[the remark this answers is not in the document]
Comment (banukutlu) [reply app/javascript/psulib_blacklight/javascripts/availability.js]: sure, will change these to camelcase

File: app/javascript/psulib_blacklight/javascripts/availability.js (hunk 4 above)
Comment (banukutlu) [reply app/javascript/psulib_blacklight/javascripts/availability.js]: > ??

changed 👍

File: app/javascript/psulib_blacklight/javascripts/availability.js (hunk 3 above)
Comment (banukutlu) [reply app/javascript/psulib_blacklight/javascripts/availability.js]: yes, I was kinda thinking about doing the same thing but forget about it later on, will do it 👍

Comment (banukutlu) [reply app/javascript/psulib_blacklight/javascripts/availability.js]: updated

File: app/javascript/psulib_blacklight/javascripts/availability.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (banukutlu) [reply app/javascript/psulib_blacklight/javascripts/availability.js]: updated

**Comment by banukutlu** [comment]
> I wonder if the expression:
> 
> `if (libraryID.toUpperCase() !== 'ONLINE') {`
> 
> could be moved up a bit to just after the line:
> 
> ` $(this).children('CallInfo').each(function () {`
> 
> It's basically a guard clause, and if the thing is ONLINE the script moves on to the next item in the `each`, so why not do that earlier?

moved the check upper 👍

File: app/javascript/psulib_blacklight/javascripts/availability.js (hunk 6 above)
Comment (cdmo) [reply app/javascript/psulib_blacklight/javascripts/availability.js]: Nice catch, thanks.

File: app/javascript/psulib_blacklight/javascripts/availability.js [hunk 8]
```diff
@@ -257,7 +257,7 @@ function groupByLibrary(holdings) {
     }, {});
 }
 
-function resolveLocation(item, illiadLocations, allLocations) {
+function printLocationHTML(item, illiadLocations, allLocations) {
```
Comment (cdmo) [inline app/javascript/psulib_blacklight/javascripts/availability.js]: Nice.

**Pull request merged**